File: block_trans/graph.py
import random
'''
用于生成和操作数据结构的单向图
'''

def gen_rule(n: int) -> dict:
    '''
    随机生成n个节点的单向图，节点编号从1到n
    图中每个节点的出度为1，出度的目标节点编号随机生成
    不能有孤立的节点，或者孤立的环
    :param n:  节点数量
    :return:   字典形式的图，key为节点编号，value为目标节点编号
    '''
    graph = {}
    nodes = list(range(1, n + 1))
    random.shuffle(nodes)
    # 选择主环的大小，从2到n
    main_loop_size = random.randint(2, n)
    # 生成主环
    for i in range(main_loop_size):
        graph[nodes[i]] = nodes[(i + 1) % main_loop_size]
    # 生成剩余的边，将剩余的节点随机连接到主环上或者已经连接到主环上的节点上
    for i in range(main_loop_size, n):
        graph[nodes[i]] = random.choice(nodes[:main_loop_size])
        main_loop_size += 1
    return graph


def gen_data(n: int) -> list:
    '''
    随机生成n个节点的数据，数据编号从1到n
    :param n:  数据数量
    :return:   数据列表
    '''
    data = list(range(1, n + 1))
    random.shuffle(data)
    return data


def transform(data: list, rule: dict, pos: int) -> list:
    '''
    把data[pos]的数据按照rule进行转换
    :param data:   数据列表
    :param rule:   转换规则
    :param pos:    点击的数字在列表data的位置
    :return:
    '''
    m = data[pos]         # 转换前的数据
    n = data[rule[pos + 1] - 1]   # 转换规则指向位置的数据
    # 交换两个数据在列表中的位置
    data[pos], data[rule[pos + 1] - 1] = n, m
    return data

File: heads/graph.cpp
#pragma once
#ifndef GRAPH_HPP
#define GRAPH_HPP

#include <algorithm>
#include <random>
#include <vector>
#include <unordered_map>

#include "graph.hpp"

std::unordered_map<int, int> gen_rule(int n) {
    /**
     * Generate a rule for a n*n game.
     * @param n: the size of the game
     * @return: a rule for the game
     */
    std::unordered_map<int, int> graph;
    std::vector<int> nodes;
    for (int i = 1; i <= n; i++) {    // init nodes = [1, 2, ..., n]
        nodes.push_back(i);
    }
    std::shuffle(nodes.begin(), nodes.end(), std::mt19937(std::random_device()()));
    // main loop size between [2, n]
    int main_loop_size = rand() % (n - 1) + 2;
    for (int i = 0; i < main_loop_size; i++) {
        graph[nodes[i]] = nodes[(i + 1) % main_loop_size];
    }
    for (int i = main_loop_size; i < n; i++) {
        graph[nodes[i]] = nodes[rand() % main_loop_size];
        main_loop_size++;
    }
    return graph;
}


std::vector<int> gen_data(int n) {
    /**
     * Generate data for a n*n game.
     * @param n: the size of the game
     * @return: a data for the game
     */
    std::vector<int> data;
    for (int i = 1; i <= n; i++) {
        data.push_back(i);
    }
    std::shuffle(data.begin(), data.end(), std::mt19937(std::random_device()()));
    return data;
}


void transform(std::vector<int> &data, std::unordered_map<int, int> &rule, int pos) {
    /**
     * Transform the data according to the rule.
     * @param data: the data of the game
     * @param rule: the rule of the game
     * @param pos: the position of the clicked number
     * @return: the transformed data
     */
    int m = data[pos];
    int n = data[rule[pos + 1] - 1];
    std::swap(data[pos], data[rule[pos + 1] - 1]);
}

bool check_win(std::vector<int> &data) {
    /**
     * data = [1, 2, 3, 4, 5, 6, 7, 8, 9 ... n] -> win
     * Check if the game is win.
     * @param data: the data of the game
     * @return: if the game is win
     */
    for (int i = 0; i < data.size(); i++) {
        if (data[i] != i + 1) {
            return false;
        }
    }
    return true;
}

#endif